### PR TITLE
Added an Error Data Listener

### PR DESCRIPTION
**Pull request by mickaelandrieu**
| Questions     | Answers
| ------------- | -------------------------------------------------------
| Branch?       | develop
| Description?  | Get information about silent errors in test suite
| Type?         | bug fix
| Category?     | TE
| BC breaks?    | no
| Deprecations? | no
| Fixed ticket? | part of https://github.com/PrestaShop/PrestaShop/projects/1#card-13082162
| How to test?  | Launch a test suite (`composer phpunit-admin` for instance)

### Exemple (See travis builds)

```
> @php -d date.timezone=UTC ./vendor/bin/phpunit -c tests/phpunit-admin.xml
PHPUnit 5.7.27 by Sebastian Bergmann and contributors.
...............................................................  63 / 114 ( 55%)
...................................................             114 / 114 (100%)

Current report of phpErrorsHandler:
Errors: 0 / Warnings: 381 / Notices: 0 / Deprecations: 159
```

---
This change is [Reviewable](https://reviewable.io/reviews/prestashop/prestashop/10477)

### CODE REVIEW

File: tests/TestCase/PhpErrorsCounter.php
Comment (mickaelandrieu) [inline tests/TestCase/PhpErrorsCounter.php:29]: ping @PierreRambaud I can throw exceptions here, but does it defeat PHPUnit configuration? :p

File: tests/TestCase/ErrorsDataListener.php
Comment (PierreRambaud) [inline tests/TestCase/ErrorsDataListener.php:57]: Could be: `PHPUnit\Framework\TestSuite`

Comment (mickaelandrieu) [reply tests/TestCase/ErrorsDataListener.php:57]: I tried but it breaks the interface

Comment (PierreRambaud) [reply tests/TestCase/ErrorsDataListener.php:57]: :/ sad
Should we plan to update our PHPUnit version?

File: tests/TestCase/PhpErrorsCounter.php
Comment (PierreRambaud) [reply tests/TestCase/PhpErrorsCounter.php:29]: We need to make some tests, with many suites, many failures, and many errors. :)

File: tests/TestCase/ErrorsDataListener.php
Comment (mickaelandrieu) [reply tests/TestCase/ErrorsDataListener.php:57]: not really important right now, let's keep that for freeze period :)

File: tests/TestCase/PhpErrorsCounter.php
Comment (mickaelandrieu) [reply tests/TestCase/PhpErrorsCounter.php:29]: Hmm write tests to test tests: no way xD

**Comment by mickaelandrieu** [comment]
Hi @PrestaShop/prestashop-core-developers, from what I see here it's already working.

Let's improve it later if needed, but it's ready for the merge isn't it? I'm thinking about the display of each type of errors according to some environment variable.

File: tests/sf-tests.xml
Comment (jolelievre) [inline tests/sf-tests.xml]: maybe we can remove the `SYMFONY_DEPRECATIONS_HELPER` then?

Comment (mickaelandrieu) [reply tests/sf-tests.xml]: Not sure it's worth it, but if we do it it should be done in another pull request because it changes the behavior.

What I do here: tracking of all PHP errors,
What the Symfony deprecations helpers do: highlight every deprecation (but mostly the Symfony one).

I'm not against making PrestaShop compatible with Symfony 4, but it will create a lot of "noise" right now when our test suite has already a lot of noise for reasons we don't even understand.

Comment (jolelievre) [reply tests/sf-tests.xml]: alright let's keep this for later

**Review by jolelievre — DISMISSED** [review]
thanks @mickaelandrieu

**Comment by jolelievre** [comment]
oh it seems some Codacy problems still need to be fixed

**Comment by mickaelandrieu** [comment]
> oh it seems some Codacy problems still need to be fixed

"Avoid unused parameters such as '$suite'." => Sadly I must respect the interface

"Expected next thing to be an escaping function (see Codex for 'Data Validation'), not 'PHP_EOL'"

I'm not sure this code will be better:

hmm

wdyt?

**Comment by jolelievre** [comment]
yes or maybe you could store PHP_EOL in a variable, or a local const? this would be smaller code
or store the double PHP_EOL
To be honest I don't understand why codacy doesn't like this
This leaves the issue about eh unused parameter `$suite`, maybe you could display the names of the suites used?

**Comment by mickaelandrieu** [comment]
> This leaves the issue about eh unused parameter $suite, maybe you could display the names of the suites used?

You mean changing the behavior of a feature to please a static code analysis tool? Hmm I'm not really easy with it.

> yes or maybe you could store PHP_EOL in a variable, or a local const?

PHP_EOL is already a constant ahahaha

I need to think a little bit about it, let's wait a little bit before merge this... maybe Codacy will be removed until then ^o^

**Comment by PierreRambaud** [comment]
If you use a concatenation you shouldn't use `printf` try:
```php
printf('%s%sCurrent report blablabla:', PHP_EOL, PHP_EOL);
```

If you don't use any arguments you can use `print` or `echo`

File: tests/TestCase/ErrorsDataListener.php
Comment (PierreRambaud) [inline tests/TestCase/ErrorsDataListener.php:71]:
```php
printf('%s%sCurrent report of phpErrorsHandler:', PHP_EOL, PHP_EOL);
printf('%s%s%s', PHP_EOL, $this->errorsCounter->displaySummary(), PHP_EOL);
```

**Comment by jolelievre** [comment]
>
> You mean changing the behavior of a feature to please a static code analysis tool? Hmm I'm not really easy with it.
> 
yeah I see your point ^^ what if you use the suite name to track the started/finished suites instead of a counter? ^^

**Comment by Quetzacoalt91** [comment]
@mickaelandrieu What's the status on this PR?

Also, I see we now have a counter, but it seems we never get details on them?

**Comment by mickaelandrieu** [comment]
> Also, I see we now have a counter, but it seems we never get details on them?

true, it's the very first step :+1:

**Comment by mickaelandrieu** [comment]
Rebased :+1:

**Review by Quetzacoalt91 — APPROVED** [review]

**Comment by Quetzacoalt91** [comment]
I don't know if we can ask a review from the QA team on this one, adding the label just in case.

**Comment by mickaelandrieu** [comment]
> I don't know if we can ask a review from the QA team on this one, adding the label just in case.

ping @marionf, if you can't ping one of the core team to merge. Even if it introduce a bug, it won't impact the customers.

**Pull request merged**
